### PR TITLE
Support arbitrary-sized tuples and arrays

### LINKED ISSUE #9: Support arbitrary-sized tuples
So far we only support pairs, not tuples with an arbitrary number of fields. That's just laziness though, and should not be hard to fix.

### PR DESCRIPTION
**Pull request by frank-king**
Fixes #9.

I think array types should be treated differently with tuples (or structs), because there are no paddings between two adjacent elements (at least for now, according to [the reference](https://doc.rust-lang.org/reference/type-layout.html#array-layout)). But I'm not sure whether it will change in the future, so I left it with a TODO.

### CODE REVIEW

**Comment by RalfJung** [comment]
That is not a good argument - tuples can have padding but they don't require padding. This makes arrays (of a fixed length) a special case of tuples.

So I am not convinced we should split arrays and tuples. Maybe this helps for place expressions (indexing vs field projection) but it does not help for (monomorphic) type representations.

**Comment by frank-king** [comment]
I didn't mean to treat arrays and tuples differently when talking about the `Type::encode` and `Type::decode` functions. I mean syntactically in MiniRust, `Type::Array` and `Type::Tuple` are different enum variants, so we should write two individual `match` arms (or fallible patterns) for them.

I just added the `Type::Array` pattern because it was missing in the original text.

To deal with `Type::Tuple` and `Type::Array` in the same match arm, we need a helper function `.fields()` which returns an iterator of offsets and field types `(Size, Type)`, and with this helper function, `Type::Tuple` and `Type::Array` can be easily processed in a single `decode` or `encode` function.

**Comment by RalfJung** [comment]
Sorry, I misunderstood (and on mobile the diff is hard to read). Will take a closer look after my vacation (in about a week).

File: lang/values.md [hunk 1]
```diff
@@ -185,27 +185,24 @@ Note that types like `&!` have no valid value: when the pointee type is uninhabi
...
-    fn decode(Tuple { fields: [field1, field2], size }: Self, bytes: List<AbstractByte>) -> Option<Value> {
+    fn decode(Tuple { fields, size }: Self, bytes: List<AbstractByte>) -> Option<Value> {
         if bytes.len() != size { throw!(); }
-        let (offset1, type1) = field1;
-        let val1 = type1.decode(bytes[offset1..][..type1.size()]);
-        let (offset2, type2) = field2;
-        let val2 = type2.decode(bytes[offset2..][..type2.size()]);
-        Value::Tuple([val1, val2])
+        Value::Tuple(
+            fields.into_iter().map(|(offset, ty)| {
```
Comment (RalfJung) [inline lang/values.md:195]:
```suggestion
            fields.iter().map(|(offset, ty)| {
```

File: lang/values.md [hunk 2]
```diff
@@ -185,27 +185,24 @@ Note that types like `&!` have no valid value: when the pointee type is uninhabi
...
-        let Value::Tuple([val1, val2]) = val else { panic!() };
-        let mut bytes = [AbstractByte::Uninit; size];
-        let (offset1, type1) = field1;
-        bytes[offset1..][..type1.size()] = type1.encode(val1);
-        let (offset2, type2) = field2;
-        bytes[offset2..][..type2.size()] = type2.encode(val2);
+    fn encode(Tuple { fields, size }: Self, val: Value) -> List<AbstractByte> {
+        let Value::Tuple(values) = val else { panic!() };
+        let mut bytes = list![AbstractByte::Uninit; size];
+        for ((offset, ty), value) in fields.into_iter().zip(values) {
```
Comment (RalfJung) [inline lang/values.md:203]:
```suggestion
        for ((offset, ty), value) in fields.iter().zip(values) {
```
also please add an assertion ensuring they have the same length

File: lang/values.md [hunk 3]
```diff
@@ -215,6 +212,32 @@ Note in particular that `decode` ignores the bytes which are before, between, or
...
+        Value::Tuple(
+            bytes.chunks(elem.size())
+                .map(|elem_bytes| elem.decode(elem_bytes))
+                .try_collect()?,
+        )
+    }
+    fn encode(Array { elem, count }: Self, val: Value) -> List<AbstractByte> {
+        let Value::Tuple(values) = val else { panic!() };
+        assert_eq!(values.len(), count);
+        values.into_iter().flat_map(|value| {
```
Comment (RalfJung) [inline lang/values.md:230]:
```suggestion
        values.iter().flat_map(|value| {
```

File: lang/values.md [hunk 4]
```diff
@@ -215,6 +212,32 @@ Note in particular that `decode` ignores the bytes which are before, between, or
...
+            let bytes = elem.encode(value);
+            assert_eq!(bytes.len(), elem.size());
+            bytes
+        }).collect()
+    }
+}
+```
+
+- TODO: Should we consider paddings between two adjacent elements in arrays?
```
Comment (RalfJung) [inline lang/values.md:239]: I don't think we need this remark, arrays are pretty tied down in how they are laid out. There is no room for padding here.

File: lang/values.md [hunk 5]
```diff
@@ -185,27 +185,24 @@ Note that types like `&!` have no valid value: when the pointee type is uninhabi
...
         if bytes.len() != size { throw!(); }
-        let (offset1, type1) = field1;
-        let val1 = type1.decode(bytes[offset1..][..type1.size()]);
-        let (offset2, type2) = field2;
-        let val2 = type2.decode(bytes[offset2..][..type2.size()]);
-        Value::Tuple([val1, val2])
+        Value::Tuple(
+            fields.into_iter().map(|(offset, ty)| {
+                ty.decode(bytes[offset..][..ty.size()])
+            }).try_collect()?,
```
Comment (RalfJung) [inline lang/values.md:197]: Can't `collect` also do this?

**Review by RalfJung — COMMENTED** [review]
Looks good, apart from a few small nits! The code duplication is small enough that a helper function is probably not worth it.

File: lang/values.md (hunk 5 above)
Comment (RalfJung) [reply lang/values.md:197]: Yeah for `Result` `collect` also works, so given that `try_collect` is unstable IMO we should use `collect` for now.

**Comment by RalfJung** [comment]
Thanks, that looks very nice!

**Pull request merged**